tests/unit/test_matrix.cu

```diff
@@ -2,64 +2,386 @@
 #include <memory/memory.hpp>
 #include <util/span.hpp>
 
-using matrix_type = nest::mc::matrix<nest::mc::gpu::backend>;
-using index_type = matrix_type::size_type;
+using namespace nest::mc;
 
-TEST(matrix, solve_gpu)
+using gpu::impl::npos;
+using util::make_span;
+using util::assign_from;
+using memory::on_gpu;
+using memory::on_host;
+
+using testing::seq_almost_eq;
+
+using std::begin;
+using std::end;
+
+// will test the flat_to_interleaved and interleaved_to_flat operations for the
+// set of matrices defined by sizes and starts.
+// Applies the interleave to the vector in values, and checks this against
+// a reference result generated using a host side reference implementation.
+// Then the interleave result is reverse_interleaved, and the result is
+// compared to the original input.
+//
+// This is implemented in a separate function to facilitate testing on a
+// broad range of BlockWidth and LoadWidth compile time parameters.
+template <typename T, typename I, int BlockWidth, int LoadWidth>
+::testing::AssertionResult test_interleave(
+        std::vector<I> sizes,
+        std::vector<I> starts,
+        std::vector<T> values,
+        int padded_size)
+{
+    auto num_mtx = sizes.size();
+
+    auto in  = on_gpu(memory::make_const_view(values));
+    auto sizes_d = on_gpu(memory::make_const_view(sizes));
+    auto starts_d = on_gpu(memory::make_const_view(starts));
+
+    int packed_size = padded_size * BlockWidth * gpu::impl::block_count(num_mtx, BlockWidth);
+
+    // forward will hold the result of the interleave operation on the GPU
+    auto forward = memory::device_vector<T>(packed_size, npos<T>());
+
+    // find the reference interleaved values using host side implementation
+    auto baseline = gpu::flat_to_interleaved(values, sizes, starts, BlockWidth, num_mtx, padded_size);
+
+    // find the interleaved values on gpu
+    gpu::flat_to_interleaved<T, I, BlockWidth, LoadWidth>(in.data(), forward.data(), sizes_d.data(), starts_d.data(), padded_size, num_mtx);
+
+    std::vector<T> result_f = assign_from(on_host(forward));
+    std::vector<T> expected = gpu::flat_to_interleaved(values, sizes, starts, BlockWidth, num_mtx, padded_size);
+    const auto forward_success = (result_f==expected);
+    if (!forward_success) {
+        return ::testing::AssertionFailure() << "interleave to flat failed: BlockWidth "
+            << BlockWidth << ", LoadWidth " << LoadWidth << "\n";
+    }
+
+    // backward will hold the result of reverse interleave on the GPU
+    auto backward = memory::device_vector<T>(values.size(), npos<T>());
+    gpu::interleaved_to_flat<T, I, BlockWidth, LoadWidth>(forward.data(), backward.data(), sizes_d.data(), starts_d.data(), padded_size, num_mtx);
+
+    std::vector<T> result_b = assign_from(on_host(backward));
+
+    // we expect that the result of the reverse permutation is the original input vector
+    const auto backward_success = (result_b==values);
+    EXPECT_TRUE(backward_success);
```

Comment:
This also should be returning a `::testing::AssertionResult` similar to line 63.